packages/mui-joy/src/styles/index.ts

```diff
@@ -71,7 +72,6 @@ export { default as styled } from './styled';
 export { default as ThemeProvider } from './ThemeProvider';
 export * from './ThemeProvider';
 export { default as useThemeProps } from './useThemeProps';
-export { sx as experimental_sx } from './styleFunctionSx';
```

Comment:
What would happen to Material UI? Do you think we should drop the `experimental_sx`?

I would be nice if the `sx` is attached to the theme directly like what I proposed in https://github.com/mui/material-ui/pull/35081/files:
```js
const Button = styled('button')(({ theme }) => theme.unstable_sx({
  px: 1,
  // ...
}))
```

Reply:
Great point, yes we can do this.

Reply:
@oliviertassinari the downside of that approach is that it is hard to debug. 

The `theme.unstable_sx(...)` is more flexible:
- You can debug the result before it goes to emotion
    ```js
    const Button = styled('button')(({ theme }) => {
      const result = theme.unstable_sx({
        px: 1,
        // ...
      });
      return result;
    })
    ```
- Since it attaches to the theme, you can use it anywhere
    ```js
    const theme = useTheme();
    const style = theme.unstable_sx({ ... });
    
    <div style={style} />
    ```
- Migration from `experimental_sx` looks easier:
    ```diff
    const Button = styled('button')(
    -  experimental_sx({ ... }),
    + ({ theme }) => theme.unstable_sx({ ... })
    )
    ```

Reply:
@siriwatknp From how I read this, the root of the question is do we want to trade less tree-shaking capabilities (see the bundle size changes) in exchange for a function easier to access (instead of an import, right from the theme). Personally, I don't really see the ROI but no objections to moving forward as is. 

> the downside of that approach is that it is hard to debug.

Is the debug experience for developers a problem with the current `sx` prop?

> Since it attaches to the theme, you can use it anywhere

What's the use case to use the function with inline style? We could also think of an alternative API like this:

```jsx
import { sxFactory } from '@mui/system';

const theme = useTheme();
const style = sxFactory(theme)({ ... });

<div style={style} />
```

> Migration from experimental_sx looks easier

I think that it's equivalent to:

```diff
const Button = styled('button')(
-  experimental_sx({ ... }),
-)
+  { ... }
+, { unstable_sxSyntax: true });
```

Reply:
> Is the debug experience for developers a problem with the current sx prop?

Definitely yes. It takes me a while to figure out why the width is '100%':

```js
<Box sx={{ width: 1 }} /> // I want to create a divider.
```

The DX here is bad because I don't know what I did wrong. However, I am not talking about `sx` here. What I mentioned was about comparing `theme.unstable_sx` and `{ unstable_sxSyntax: true }`.

With your proposal, logic lives inside the `styled` which takes time for developers to figure out what's wrong with their code. I think the whole purpose of this PR is to move the logic to the theme to make it more flexible.

But, with `theme.unstable_sx`, it is just a function and more flexible. I can debug the result of it then I will see that this function somehow turns the `width: 100%`. The feedback loop is very short.

Reply:
> I think that it's equivalent to:

Not for the case that simple styles are mixed with `experimental_sx`:

```js
const Button = styled('button')(({ theme }) => ([
  { padding: 8 },
  experimental_sx({
    '&:hover': {
      bgcolor: 'primary.main',
    }
  })
]))
```

Adding `{ unstable_sxSyntax: true }` will turn `padding: 8` to `padding theme.spacing(8)`.

Reply:
@siriwatknp Right, ok. Then I think that it could be valuable to consider this alternative API. Maybe it can help with tree shaking and doesn't harm the DX too much:

```jsx
import { styled, unstable_sx } from '@mui/system';

const Button = styled('button')(({ theme }) => ([
  { padding: 8 },
  unstable_sx(theme, {
    '&:hover': {
      bgcolor: 'primary.main',
    },
  })
]));
```

> It takes me a while to figure out why the width is '100%':

I think that it could be great to make a breaking change, to remove 1 === 100%. +1 to open a GitHub issue with this one. I had the same frustration in the past.

Reply:
> tree shaking

How would it help when `unstable_sx` is used internally by the `sx` prop?

Reply:
https://mui-dashboard.netlify.app/size-comparison?circleCIBuildNumber=454127&baseRef=master&baseCommit=b8b1e0d08ceab826a3e11b18a3b4b6e6b5c17206&prNumber=35150 answers this question, but I agree that to fully benefit from this, we would need to change the default option of `createStyled` in the system to not come with the sx helpers by default. Not sure if it's worth it.